CAM Helix, add feedrate radial correction and improve retract()

J-Dunn opened the pull request:
Incorporates changes to helix retract() previously reviewed and discussed in: 
CAM: Improved retraction between helical paths #20340

Adds new dlg option for radial adjustment of spindle feed rate parameter "F" 

As radius changes from on helix to the next in multiple cut helix operations the spindle feedrate is adjusted for the radial difference between cutting edge of tool and spindle path radius. 

The aim is to give roughly constant tool cutting speed across all helices, allowing one ToolController
Spindle feedrate is reduced when cutting outer edge and increase when cutting inwards to maintain tool tip speed. 

The effect can be a ratio of 2 or more when tool diam is near hole diameter. 
A cutoff is set at 200 to avoid div zero condition.

The issue is explained here if required. 
https://www.youtube.com/watch?v=6wLU97gVo5k

phaseloop commented on src/Mod/CAM/Gui/Resources/panels/PageOpHelixEdit.ui:
```diff
@@ -161,7 +159,17 @@
+   <item row="2" column="0">
+    <widget class="QCheckBox" name="FeedRateCheckBox">
+     <property name="toolTip">
+      <string>Radial feed rate correction</string>
```
Would be great to add here a detailed description - the one you provided in PR about keeping feedrate constant with different helix radius :)

phaseloop commented on src/Mod/CAM/Path/Base/FeedRate.py:
```diff
@@ -57,26 +58,47 @@ def _isVertical(currentposition, command):
         return Path.Geom.isVertical(Part.makeLine(currentposition, endpoint))
 
+    def _isHorizontal(currentposition, command):
+        x = command.Parameters["X"] if "X" in command.Parameters else currentposition.x
```
python protip:

`command.Parameters.get("X", currentposition.x)`  - will get "X" or use second argument if dictionary key is not found :)

J-Dunn replied:
I just copied that code from isVertical() which is long established. 
I suppose we could replace all 6 occurrences of the confusing python if assignment syntax with an even less readable equivalent.   I'll consider whether that's any better. 
Thanks for the tip.

phaseloop replied:
Nah it's just a tip. If you want to keep current version, it's fine as well.

phaseloop commented on src/Mod/CAM/Path/Base/Generator/helix.py:
```diff
@@ -148,6 +161,27 @@ def helix_cut_r(r):
+                    ) / r  # startAt outside: increase spindle feed rate to maintain inner-cut chip-load
+
+        # Note: if (r == -tool_radius), helix is plugne, hence VH_gradient==0; feedrate==vFeed , OK
+        print("rad = ", r, ",  feedRateRatio = ", feedRateRatio)
```
debugging leftover? If not - Path.Log.debug should be used to not spam console on production :)

phaseloop replied:
also to make your life easier (or more confusing python syntax, depending on perspective, lol):
you can use syntax like that (with `f` marking the beginning of string)

`print(f"rad={r}, feedRateRatio={feedRateRatio}})`

J-Dunn replied:
Ah, unmatched quotes, what's not to like. Or was that  a typo?

phaseloop replied:
a typo, fixed it

phaseloop commented on src/Mod/CAM/Path/Base/FeedRate.py:
```diff
@@ -57,26 +58,47 @@ def _isVertical(currentposition, command):
+        params = command.Parameters
+        if ("F" in params) and (command.Name in ["G2", "G3"]):
+            # adj. spindle feedrate to ensure correct chip load on cutting edge, if "F" defined
+            hFeed_adj = hFeed * params["F"]
```
I guess that you are piggybacking on `F` parameter to pass the ratio, right? Then this is converted to real F value
using tool speed from configuration.

As Parameters is just a dictionary I think it would be better to just use literally a key "FeedRateAdj" (or similar).

The issue is - something goes wrong or another developer introduces a bug in the code - and some random `F` value (or `0`) lands in resulting g-code damaging the machine/bit/stock.

J-Dunn commented:
Very good comments. I'll make few update when I have time later.

phaseloop commented on src/Mod/CAM/Path/Base/Generator/helix.py:
```diff
@@ -115,26 +126,28 @@ def generate(
+        outer_path_radius = hole_radius - tool_radius
+        inner_path_radius = inner_radius + tool_radius
+
+        if abs((outer_path_radius - inner_path_radius) / step_over_distance) < 1e-5:
```
Instead of constant you may want to  use tolerance set for particular job?

https://github.com/FreeCAD/FreeCAD/blob/b3d8020e7777b1f6c9dcd5ebfc10c967b3d02485/src/Mod/CAM/Path/Preferences.py#L120

`def defaultGeometryTolerance()`

J-Dunn replied:
I agree in principal but look again, it's a ratio not a tolerance.

phaseloop replied:
AFAIK it would make sense to just compare `abs(outer_path_radius-inner_path_radius) < job_tolerance` instead of ratio here.

Also why this particular value? If the ratio is slightly larger than 1e-5, next calculation at line 147:

```
nr = int(ceil((outer_path_radius - inner_path_radius) / step_over_distance) + 1)
```

will compute to nr=2, so in line 148 in linear space we will have two steps with some microscopic radius difference like `0.00006` mm

J-Dunn replied:
You have a basically valid point. That was not my code, the test was there already in some form but when true it tried to split the difference, which was pointless. If I can rapidly find valid tolerance variable, I'll use it. Otherwise it's rather academic.

phaseloop replied:
Here you have an example:

https://github.com/FreeCAD/FreeCAD/blob/b3d8020e7777b1f6c9dcd5ebfc10c967b3d02485/src/Mod/CAM/Path/Op/Vcarve.py#L343

Would be great if you could fix it if you have this PR open already - CAM workbench if full of bugs and weird behaviours.

J-Dunn replied:
OK, can you tell what variable I need I can't get this to fly and I can't spend hours looking.

phaseloop replied:
what do you mean? This is how you retrieve default job tolerance:
```
obj.Tolerance = Path.Preferences.defaultGeometryTolerance() 
```

J-Dunn replied:
Thanks for the tip.  obj.Tolerance did not exist when I tried that. I did not want to stamp on an existing value with a default if something else sets it.  I'll bow to your expertise. 

The following code works.
`if abs(outer_path_radius-inner_path_radius) < Path.Preferences.defaultGeometryTolerance() :`

Do you suggest I add that the PR?

phaseloop replied:
yes, the one you posted is fine - thanks a lot!

phaseloop commented:
Can't comment much about math behind it but overall seems fine - left few comments, AFAIK most important issue from my perspective is to not piggyback on `F` but use something custom.

J-Dunn commented:
" left few comments"
Do you see more than the one you pointed out?

phaseloop commented:
> " left few comments" Do you see more than the one you pointed out?

- inner/outer radius tolerance to discuss (using job tolerance setting)
- Adding longer tooltip description in UI widget
- `print()` statement left - either as accident or should be converted to log.debug call

this + python tips = "few" :D

J-Dunn commented:
I beg your pardon, I thought you were saying there where other forgotten debugging print comments.

phaseloop commented:
> I beg your pardon, I thought you were saying there where other forgotten debugging print comments.

It should have been "I left few comments" - in PR review. My bad.

J-Dunn commented:
all good now. 
Ready to merge ?

phaseloop commented:
Fine for me - you need a blessing from someone with merge permissions too

dbtayl commented:
A couple meta-comments...

I'd recommend setting up pre-commit to avoid the "auto fixes" commits. Way easier to catch those before you commit than having to go fix commits later or squashing auto-fixes that get committed. See here: https://freecad.github.io/DevelopersHandbook/gettingstarted/#setting-up-for-development

I'd also suggest squashing your commits. See here (item 7): https://freecad.github.io/DevelopersHandbook/gettingstarted/#pr-review-process Just going by the messages on the commits in this PR, it looks like there's a lot of going back and changing/correcting previous commits. All that doesn't need to go in the PR, just the clean final result.

J-Dunn commented:
> Just going by the messages on the commits in this PR...

Well if you took the time to read the above comments before commenting, you would see those changes were in response to feedback and comments for reviewers, not my uncertainty or changing my mind. 

Thanks for pointing out pre-commit,  I was not aware that was available as a distro package. I'll check it out.

J-Dunn commented:
@sliptonic 
@Russ4262 

Hi, you two seem to be heart of this code. Maybe you like to look it over. 

The main change introduced in adjusting feedrate is very short: 

helix.py: 165

```
        #FeedRateCheckbox.checked
        if not feedRateAdj : feedRateRatio = 0
        else:
            if (startAt=="Inside") or (r == radii[0]): # first cut, always adj for outer wall chip-load
                feedRateRatio = r/(r+tool_radius)  # classic (hole_rad-tool_rad)/hole_rad adjustment
            else:
                if tool_radius > 200*r : feedRateRatio = 200 # prevent div zero
                else: feedRateRatio = (r+tool_radius)/r # startAt outside: increase spindle feed rate to maintain inner-cut chip-load

```
Feedrate.py : 79
```
        params = command.Parameters
        if ("FeedFactor" in params) and (command.Name in ["G2","G3"]):
            # adj. spindle feedrate to ensure correct chip load on cutting edge, if "F" defined
            hFeed_adj = hFeed* params["FeedFactor"]
            rate = math.sqrt(vFeed * vFeed + hFeed_adj * hFeed_adj) # vector sum
            # TODO combined V,H feedrates for 3D-ops, ramp, pocket/contour radius ?

```
The retraction code was previously reviewed in #20340 and was included here without further change. 

Thanks.

dbtayl commented:
> Well if you took the time to read the above comments before commenting, you would see those changes were in response to feedback and comments for reviewers, not my uncertainty or changing my mind.

Please don't take what I've said personally- I saw what looked like a knowledge gap in the FreeCAD development process and figured you might like to be aware of the guidelines the project has in place to more effectively contribute, since it's not always easy to find the right documentation. I have no authority with the project, so if you feel my comments are incorrect, ignoring them costs you nothing.

That said, I _did_ read the comments, just not your code. My point is from the perspective of somebody touching the FreeCAD codebase in the future, having all of the intermediate steps partially implementing a feature is actively unhelpful. It's _way_ easier to see a single commit that implements the feature with message (eg) "CAM: Helix, add feedrate radial correction and improve retract" than a _bunch_ of commits with messages like "add feedrate adjustment checkbox", with zero context re (eg) what operation it applies to or what set of commits constitutes the feature.

Might also help get pull requests approved faster- seems plausible it's easier to (or folks are more inclined to) review a single, nicely-formatted commit than a whole development history of a feature. Lower activation energy.

J-Dunn commented:
Thanks for your reply. 
You prefaced your comments with "Just going by the messages on the commits in this PR..."  I took that a face value. 
It now seems that your main point was that this should be one commit, not one commit for each file change. That was not clear before from the way you presented it. 
You may have point there. That is a result of my lack of experience with github. I thought I had to commit each file change, so added a comment for each one. 
Having said that, if I was reviewing a commit I'd rather have 50 char comments like  "add feedrate adjustment checkbox" etc. so I can see which commits are the core code changes which need attention, and which are just correcting silly slip-ups like calling the hole diameter variable  hole["r"]. ;)

I originally split off the retract change since it was totally independent and I thought it may help to review smaller, single function changes. However, that sat without getting merged for two weeks, so I bundled it into one PR.

I appreciate your link to the guide. These things are often not easily to find. Even the build instructions are not linked anywhere on the main site or downloads section. 

I certain don't take any of it personally. Have a nice day.

phaseloop commented:
Git is unfriendly, so I'll share my method of doing that:

- use "git log" to get the commit hash of the last commit that is NOT yours (e.g the one preceding your first commit)
- git rebase -i <hash from above>
- on the list of all your commits, leave first one untouched and modify all other lines to have "s" (squash command) - there should be a manual displayed in the same file you edit
- save the file, another editor should pop-up where you edit commit message. Delete/comment all lines and leave only the one that should be the real commit message
- save the file
- git push -f to push changes to github. You need -f (force) flag because as you rebased your commits, github will reject those changes as not valid. If you are using VS Code do not let it "sync" changes, just force push it.

J-Dunn commented:
Thanks for the tips. 
This one will stay the way it is now. Unless necessary coding changes up.
I'll try to be more concise if I do another PR.

dbtayl commented:
> This one will stay the way it is now. Unless necessary coding changes turn up. I'll try to be more concise if I do another PR.

> those changes were in response to feedback and comments for reviewers

This is _exactly_ why the squash is the last step before merging- making it nice and tidy _after_ all the feedback has been incorporated.

That said, turns out I've been a bit misleading here- I learned something today. Looks whoever merges the PR can check a box to squash things to a single commit prior to merging. If there's a reason your PR _should_ be multiple commits (probably not this one), you'd still want to squash it yourself down to a minimal set of self-contained commits, but that's unecessary for something that's a single commit anyway.

It "looks" tidier if you squash it anyway, but from a purely technical perspective it doesn't appear to matter in this case.

J-Dunn commented:
Thanks for the update sliptonic. 

Have you looked at the PR?

sliptonic commented:
> Thanks for the update sliptonic.
> 
> Apart from the technicalities of GH, have you looked at the PR?

Looking at it today.  Also, I just learned that we're not using the merge queue so maintainers can do the squash/merge again.

sliptonic commented:
I watched the linked youtube video and I think I understand the intent of the PR.  The result doesn't look correct though.
In this screenshot, the left side is the helix op with output and radial correction enabled.   On the right is the 1.0 build with the same helix.  It looks like the feedrate is FASTER than expected.  Are you sure you applied the corrected feedrate algorithm correctly?

![image](https://github.com/user-attachments/assets/55aea010-c066-4d75-9633-4538d1dbc673)

I think this is a good improvement.  Ideally, the checkbox would not be on the helix operation but on the tool controller and the feedrate correction would be applied to all inside arc moves.  That way the same improvements wouldn't have to be made on every operation.  Per the video, this is particularly troublesome for threadmilling.

J-Dunn commented:
" It looks like the feedrate is FASTER than expected. "

For some reason I can't access the full size of the screen shot, but this cuts both ways (no pun intended). 
If it is the inner side of the tool cutting, you need speed up the spindle feed rate to maintain tool tip speed, and slow it down if it's the outside cutting. 

You can see how the two cases are done in the code snippet I posted above. 

Yes, ideally this would be applied to all arc path elements and I designed it in way that this could be easily extended. 

Unfortunately it does need some knowledge of the context so the FeedRatio needs to be created in the  code for each path operation. If that is done , it can be applied with generality in  feedRate.py as done here. 

I had a quick look at how this could be applied to contour/profile but quickly ended up in the inpenetrable jungle of the libArea code ;) 

I think it would be more feasible to apply it to thread milling which basically a helix set up wizard. 

If it ever does get to be universally applied the checkbox could move could be considered. For now I think it has to be on those ops which support it.  

I'm currently adding the "Start Radius" to the UI. This is functionally too useful to left quietly out of view. I'd used this a lot and was never aware it was there. 

Thanks for taking a look.

sliptonic commented:
In the video, he's starting with F30 and compensating to end up at F2.7 (Elevent times slower)
(https://youtu.be/6wLU97gVo5k?si=5lcTtPOxY6_1ok_u&t=282) .  

In my example, I'm setting a Feedrate of 100 mm/min and compensating the inside cut which results in a Feed rate of 111.  (10% faster).   We need to slow down the programmed feedrate because the feedrate at the edge of the tool is too fast.  Instead, it looks like we're speeding up the programmed feedrate.

What am I missing?

J-Dunn commented:
In his example he's cutting the outer edge with a tool a little smaller than the hole (this is where the errors can be huge).
If I follow you description you are cutting the inside , ie you did a startAt=="Outside" path.  In that case you need to run around the central post with a spindle feed similarly faster to maintain tool tip speed. 

If you could post a FCStd, I'd be able to answer with more certainty and make sure you have not found an error.

J-Dunn commented:
Ah thanks, code and data is always clearer than words. 

There are a couple of things here since this operation clears in one path. Again this is shown with comments in the snippet I posted where I pinged you and Russ. 

The first cut is always optimised for the outer cutting edge , since this gets the most chip load and is cutting with full tool width.  In this case startAt  Inside or Outside is the same. 

If you look at where the spindle feedrate is calculated ( also above ):
`rate = math.sqrt(vFeed * vFeed + hFeed_adj * hFeed_adj) # vector sum`

You specified h=100; v=100 ; your tool,hole radii give a feedrate adjustment 0.5 (reduced as you expected). 
sqrt(100*100+50*50)  = 111.803

So at least the code is doing what was intended. 

This does it's best to follow user's chosen v,h feedrates : when the controller converts the arc into micro dx,dy,dz movements (in the case of a classic orthogonal CNC),  instantaneous H or V feedrates you should get back to 50,100 ie the vfeed you specified plus the derated hfeed.  Since this issue only concerns the horizontal plane vfeed is not altered. 

This expects that the machinist will choose appropriate feed rates for his tool when cutting a helix.  If this is different from how he would use the same tool for a different kind of path, that is his responsibility to set up a different TC. 

So I think the bit you were missing is that a helical arc is a three dimensional movement and the feedrate is the instantaneous tangential feedrate, not just the horizontal component.   This is obviously quite significant in a steep helix, as in your example test file. I hope that makes more sense.

sliptonic commented:
> > It's only done in Drilling/Helix/Tapping now
> 
> Oh, I guess I'd assumed it was done the same throughout. At least it should be easy to extend to threadmilling.

I implemented the FeedRate helper when I started implementing the generators.  We only have those for Drilling/Helix/TM so far.  


> > You wouldn't need to worry about the nastiness in libarea/patharea guts.
> 
> Sadly, I doubt that is correct. Firstly we need to know whether it's inside or outside. And as you see with Helix there are quite a few ifs and buts which need to be considered. I spent 4 days once trying to make sense of libArea and gave up on it.

Yeah, it really needs work.  At a bare minimum it needs to be using the latest clipper but with the PathArea wrapper, that's a big job.   Ideally, we ditch the whole thing in favor a new offsetter like Cavalier Contours.  Unfortunately, that's written in Rust and we don't have things in place to use Rust code yet.

> For the moment I'm trying to see what is wrong with extraOffset , it seems that any change done in the dlg does not make it back to the object. I posted on the forum about this but no replies yet.
> 
> Do you know how the values are set and read in getFields() and setFields() ? Please post to the forum if you have some idea on that.

It's part of the operation class hierarchy.  Basically it's all done automatically which allows for Job Templates and setup sheets.  I can't remember all the details to get it working.  I always pick another property name and grep my way upstream until I figure it out.  Sorry I'm not more help there.  BTW,  I stopped posting on the forum a long time ago.  I only participate in the FPA sections now.

J-Dunn commented:
>  I always pick another property name and grep my way upstream until I figure it out. 

Yes, that's what I did to add FeedRateAdj and it went fine.  Here I tried to copy ExtraOffset until I discovered that was not working correctly to start with. https://forum.freecad.org/viewtopic.php?t=96160

It may be a while since I used that feature but I'm sure I would have done it from UI and I'm sure it used to work.
Is this a Qt6, Pyside6, shibroken regression ?

J-Dunn commented:
The StartRadius property is very confusing in the current context of what Helix can do.
I may have made sense when this was conceived but when user selects StartAt="Outside" it becomes contradictory.

Obviously since these are not just internal code varialbe names but automatically exposed to the user in the properties table , the names matter. 

Maybe this needs to be replaced and renamed with a hasAttr test in ondocumentload().

Is it possible to change this without breaking scripts and macros a making a mess ?

phaseloop commented:
I suppose major issue is backward compatibility with older FC versions and setup sheets. I suppose you can keep startRadius attribute and just make sure it is synced with innerRadius on each onDocumentLoad and exec(). I don't think any script or macro accesses attributes directly so we should not worry about it.

J-Dunn commented:
> It's only done in Drilling/Helix/Tapping now, but could probably just be added to the end of the base op class (I haven't looked yet).

I have been able to add the same thing to Threadmilling ops, though that still needs some testing before I add a PR. 

It may be best to merge this one for the Helix paths and I'll create another for Threadmilling. 
There's probably too much to lump together and smaller chunks make it easier if there is ever any regression found.

sliptonic commented:
> 
> Personally I don't have time to take this any further. I'll have a look at adding some more specific tests, that should not take too long.

Let's get some test of the new algorithm in and then merge

hyarion commented on src/Mod/CAM/Path/Base/Generator/helix.py:
```diff
@@ -148,6 +163,26 @@ def helix_cut_r(r):
         commandlist.append(Path.Command("G1", {"Z": startPoint.z}))
+
+        # FeedRateCheckbox.checked
+        if not feedRateAdj:
+            feedRateRatio = 0
+        else:
+            if (startAt == "Inside") or (
+                r == radii[0]
+            ):  # first cut, always adj for outer wall chip-load
+                feedRateRatio = r / (
+                    r + tool_radius
+                )  # classic (hole_rad-tool_rad)/hole_rad adjustment
+            else:
+                if tool_radius > 200 * r:
+                    feedRateRatio = 200  # prevent div zero
+                else:
+                    feedRateRatio = (
+                        r + tool_radius
+                    ) / r  # startAt outside: increase spindle feed rate to maintain inner-cut chip-load
```
This would also benefit of using `elif`:

```suggestion
        if not feedRateAdj:
            feedRateRatio = 0
        elif (startAt == "Inside") or (r == radii[0]):
            # first cut, always adj for outer wall chip-load
            # classic (hole_rad-tool_rad)/hole_rad adjustment
            feedRateRatio = r / (r + tool_radius)
        elif tool_radius > 200 * r:
            feedRateRatio = 200  # prevent div zero
        else:
            # startAt outside: increase spindle feed rate to maintain inner-cut chip-load
            feedRateRatio = (r + tool_radius) / r
```

chennes replied:
@J-Dunn - participation in Open Source development requires a spirit of cooperation and good will. We ask that *any time* you are working with a reviewer on a PR their comments be taken with that spirit in mind. Deeply-nested code represents a significant cognitive load and immediate "code smell", and @hyarion's comment is not unreasonable. Please bear the [Code of Conduct](https://github.com/FreeCAD/FreeCAD/blob/main/CODE_OF_CONDUCT.md) in mind as you craft your responses to reviewer comments and keep them respectful, considerate, and forgiving.

J-Dunn replied:
Anyone experiencing a significant cognitive load  when seeing deeply-nested code should probably take a break from auditing code. 

Personally I find indented blocks immediately  clear in structure in the bat of an eye. Non-indented elifs are nice if you have line length problems but they do require reading and mental parsing of the written code. 

That may be the way my mental processes work but that is why indenting was invented before python decided to make  an obligatory part the syntax rules: it is clearer and easier to read and interpret than non-indented code.

J-Dunn commented on src/Mod/CAM/Path/Base/Generator/helix.py:
```diff
@@ -115,26 +126,30 @@ def generate(
-        step_radius = inner_radius + tool_diameter / 2
-        if abs((outer_radius - step_radius) / step_over_distance) < 1e-5:
-            radii = [(outer_radius + inner_radius) / 2]
+    else:  # inner_radius >= 0:
```
> The if checks inner_radius <= 0 so the else isn't called when it is equal to 0.
Where do you see that ?

hyarion commented on src/Mod/CAM/Path/Base/Generator/helix.py:
```diff
@@ -115,26 +126,30 @@ def generate(
+                "Cannot creatre a helical path of diameter {0} with a tool of diameter {1}".format(
                     2 * hole_radius, tool_diameter
                 )
             )
-        outer_radius = hole_radius
-
-    else:  # inner_radius > 0:
-        Path.Log.debug("(annulus mode / full hole)\n")
-        outer_radius = hole_radius - tool_diameter / 2
-        step_radius = inner_radius + tool_diameter / 2
-        if abs((outer_radius - step_radius) / step_over_distance) < 1e-5:
-            radii = [(outer_radius + inner_radius) / 2]
+    else:  # inner_radius >= 0:
```
Just a note from the other, now resolved, comment:
Comment on [`else`](https://github.com/FreeCAD/FreeCAD/pull/20537/files#diff-5735e1231038eb2ebfffeb80742a7dddda7c809600e72a1ea54ed044e8530612R139) (`# inner_radius >= 0:`) doesn't match the [`if`](https://github.com/FreeCAD/FreeCAD/pull/20537/files#diff-5735e1231038eb2ebfffeb80742a7dddda7c809600e72a1ea54ed044e8530612R130) (`if hole_radius <= tool_diameter:`).

J-Dunn replied:
If the other note is resolved why do you relaunch the issue again?  Fortunately you've  discovered how to include the relevant "if" statement using GH. 

As you will note this erroneous comment is already in master.  All I did was add the "=".  So the issue about it not corresponding to the if clause currently still a live issue.  Since it seems to be of significant importance to you maybe you would like to create a PR to remove it entirely.

hyarion replied:
> If the other note is resolved why do you relaunch the issue again? Fortunately you've discovered how to include the relevant "if" statement using GH.

You where unhappy that I didn't include the `if` in the last comment so I replaced it with this new one ("resolved" the old one) as it isn't possible to change scope on github.

> All I did was add the "=".

Before your change it was probably just missed, now you've incorrectly "fixed" it. I don't see it unreasonable to ask you to properly fix the issue, or drop your change if you really don't want to make it correct.

J-Dunn replied:
I was not unhappy that you didn't include the if, I was unhappy that you had not even bothered reading the code and made false statements about what the code actually contained.  I pointed out that if you had taken the trouble to find and quote the relevant if , your mistake would have been obvious to you and you would not have made false claims. 

If you are going to review code , you need to actually read it first. 

I've now removed the offending comment entirely, so try to move on.

sliptonic commented:
Reviewed at merge meeting today.  Hold pending unit tests

J-Dunn commented:
> Reviewed at merge meeting today. Hold pending unit tests

Thanks.  I have added some fairly comprehensive units tests for the new features and extended the existing ones. 

In testing I realised that it is necessary to pass both the horizontal feedrate adjustment and a parameter describing the vertical gradient of the helix, in order to appropriately blend V,H feedrates.  (Current code is a crude "isVertical" test, all else getting just horizontal feed irrespectively).

I have used the magnitude of the vector sum of V and H (adjusted) for an initial F value. Then project the resultant using the helix gradient and clip to ensure that neither direction exceeds the intended feedrate. 

I have modified Threadmilling to use the same feedrate adj. and pass through FeedRate.py to set F , as is done with Helix. This nicely unifies the two "circularHole" ops. 

I need to do some more checking on threadmilling to check the upward cutting is handled correctly.

J-Dunn commented:
In my previous PR on this , it was recommended not to temporarily piggy-back "F" with the new "FEEDFACTOR" temporary param. 

However, now I have worked around some CommandImp.cpp oddity which was masking these values, FC is complaining when restoring Helix.nc  on reopening such a file.  It seems to be barfing on finding non-std gcode params (sensible).

Should I modify the file read-in code or go back to piggy backing existing params which do not cause issues?

sliptonic commented:
> Should I modify the file read-in code or go back to piggy backing existing params which do not cause issues?

To be honest, I'm having a hard time keeping up.   48 comments on the previous PR.  67 (and counting here).  Parts of the conversation left unresolved.  It's a bit overwhelming.

I searched the previous PR looking for a discussion of 'Piggy backing' and came up empty.  I don't understand how your algorithm works so I'm struggling to be helpful.

Can you give a simpler description of what you're trying to achieve with the command and temporary params?

J-Dunn commented:
I agree there has been a lot of irrelevant time wasting comments on this PR. That's a distraction. 

To calculate the corrected feeds in FeedRate.py I need to pass two bits of information from the generator. 
1. A ration, based on path and tool radii
2. The descent rate of the helix. 

Since "F" is not set at that stage I was putting the ratio in "F" and setFeedRate() used this to create the real "F" and sets the parameter value. That seemed harmless but some suggested it could cause confusion.  I hope that's a little clearer. 

Due to the buggy behaviour of setParameters() these were being magically hidden and it worked fine. Avoid that bug seems to cause problems when they are present and correct. I need to dig into that to confirm.

sliptonic commented:
I didn't say they were irrelevant or time wasting.  I said there were a lot of them.  Let's give the benefit of the doubt that everyone wants a good long-term solution and is operating in good faith.

Without studying too deeply, it looks like the ratio is consistent for the entire path, not calculated per-segment.  So can't you pass an optional ratio parameter to the feedrate helper?

J-Dunn commented:
suggestion: 
1. only single letters are defined and all letters are defined for something. However, UVW are clones of ABC, and FreeCAD decodes ABC.  In essence UVW are "free". 
2. No two letter codes are define, so use DR and FF and mod  setFromGCode() to handle double letter keys. 

Warming, option 2 may require finding someone with a PhD is multi-dimension non-linear system dynamics to unravel the twelve levels of nested if statement in the current code to add another layer or two !

J-Dunn commented:
cleaned all this up into new PR
#21041
Now added same to ThreadMill paths, integrated in the FeedRate.py
test units for Helix and Threadmill , as suggested. 

Closing rather lengthy and mess one.

hyarion commented:
@J-Dunn Let me know if you want help pushing your new changes here.

J-Dunn commented:
> @J-Dunn Let me know if you want help pushing your new changes here.

No thanks, I've had enough of your "help".
